core/packages/test/scripts/configure-statemine.sh

```diff
@@ -138,42 +59,103 @@ add_exporter_config() {
 
-    messages:
-        add-exporter-config - adds bridge config for ethereum.
-EOF
-    exit 1
-fi
+    local junction=$(jq --null-input \
+                        --arg eth_chain_id "$eth_chain_id" \
+                        '{ "GlobalConsensus": { "Ethereum": { "chainId": $eth_chain_id } } }')
 
-while [ "$#" -gt 0 ]; do
-  case "$1" in
-    add-exporter-config)
-      shift
-      add_exporter_config $1 $2 $3 $4 $5 $6
-      shift 6
-      ;;
-    *)
-      echo "Unknown message: $1"
-      exit 1
-      ;;
-  esac
-done
-
-exit 0
+    echo ""
+    echo "          location:"
+    echo "${location}"
+    echo ""
+    echo "          junction:"
+    echo "${junction}"
+    echo ""
+    echo "--------------------------------------------------"
+
+    local tmp_output_file=$(mktemp)
+    generate_hex_encoded_call_data "add-universal-alias" "${statemine_ws_url}" "${tmp_output_file}" "$location" "$junction"
+    local hex_encoded_data=$(cat $tmp_output_file)
+    rm $tmp_output_file
+
+    send_governance_transact_from_relaychain "${statemine_para_id}" "${hex_encoded_data}" 200000000 12000
+}
+
+add_reserve_location() {
+    echo "  calling add_reserve_location:"
+
+    local nativeTokens=$1
+    # Ethereum Native Tokens contract
+    local reserve_location=$(jq --null-input \
+                        --arg eth_chain_id "$eth_chain_id" \
+                        --arg nativeTokens "$nativeTokens" \
+                        '{ "V3": {
+                            "parents": 2,
+                            "interior": {
+                                "X2": [
+                                    {
+                                        "GlobalConsensus": { "Ethereum": { "chainId": $eth_chain_id } },
+                                    },
+                                    {
+                                        "AccountKey20": { "network": { "Ethereum": { "chainId": $eth_chain_id } }, "key": $nativeTokens }
+                                    }
+                                ]
+                            }
+                        } }')
+
+    echo ""
+    echo "          reserve_location:"
+    echo "${reserve_location}"
+    echo ""
+    echo "--------------------------------------------------"
+
+    local tmp_output_file=$(mktemp)
+    generate_hex_encoded_call_data "add-reserve-location" "${statemine_ws_url}" "${tmp_output_file}" "$reserve_location"
+    local hex_encoded_data=$(cat $tmp_output_file)
+    rm $tmp_output_file
+
+    send_governance_transact_from_relaychain "${statemine_para_id}" "${hex_encoded_data}" 200000000 12000
+}
+
+configure_statemine() {
+    # Allows messages to be sent from Substrate to ETH.
+    add_exporter_config
+    # Allow messages to be sent from ETH to Substrate.
+    add_universal_alias
+    # Allow NativeTokens contract to ReserveAssetDeposited xcm instruction.
+    add_reserve_location $(address_for NativeTokens)
```

Comment:
We could also use vault contracts as a reserve location as they live forever and actually hold funds. So maybe a better choice than a dummy address.

Reply:
> I was thinking that for upgrades we would just add the new NativeTokens contract address as a reserve location using the same `add_reserve_location` extrinsic, potentially even in the same call which sends the upgrade message across the bridge. The `bridge-transfer` pallet can have both versions of the contract added and can be run back to back. Once upgrade is complete we can remove the old contract address using `remove_reserve_location` extrinsic.
> 
> Do you see any problems with this approach?

There are various problem with this approach.

It makes smart contract upgrades, slower, more brittle. Since it requires two governance proposals: (1) Update the ethereum side, (2) update the polkadot side. Alternatively, if you combine the proposals, then if one subtask fails, you end up with inconsistent state which requires more governance proposals to fix.

Remember that it can take weeks to execute a governance proposal.

Reply:
> We could also use vault contracts as a reserve location as they live forever and actually hold funds. So maybe a better choice than a dummy address.

Vault contracts are in internal implementation detail. I don't want to expose them in our public API, which could cause problems if ever change our design/implementation later down the line.

Reply:
Ah I see. I have created this issue to switch to a dummy address.
https://linear.app/snowfork/issue/SNO-450
I think we should defer this for now. While I am focusing on trying to get the round trip transfer done we can continue with using NativeTokens?

Reply:
It won't work without a dummy address, and we need to update BridgeHub so that the XCM messages it sends to Statemint are seen as coming from a MultiLocation containing the dummy address as a junction.

Because the XCM config in Statemint/Westmint has this constraint:
https://github.com/paritytech/cumulus/pull/2013/files#diff-07287c62dd91150d975f7504bd14cc62d9889c44f7ba8ca44c30c03b05d5b071R383

https://github.com/paritytech/cumulus/blob/c167adde899a1a3818e19cf55dfe9a6b907bff63/parachains/pallets/bridge-transfer/src/impls.rs#L43

It checks that an asset has a multilocation that has the reserve location as a prefix.

Reply:
> If we replace NativeTokens with an upgraded contract, that will break the config on Statemint.

 Seems proxy pattern will fit well into this case, the proxy address of `NativeTokens` will not change when underlying logic contract upgraded. Another benifit with this pattern is storage isolation so no need for migration when upgrading.
 
 Cons are as Vincent mentioned before a bit brittle and not easy to maintain. https://github.com/Snowfork/snowbridge/pull/801#discussion_r1162432579